syntax = "proto3";
package EloqDS.remote;
option cc_generic_services = true;
import "google/protobuf/empty.proto";

service DataStoreRpcService {
  rpc FetchRecords(FetchRecordsRequest) returns (FetchRecordsResponse);
  rpc BatchWriteRecords(BatchWriteRecordsRequest)
      returns (BatchWriteRecordsResponse);
  rpc CkptEnd(CkptEndRequest) returns (CkptEndResponse);
  rpc FetchTableCatalog(FetchTableCatalogRequest)
      returns (FetchTableCatalogResponse);
  rpc UpsertTable(UpsertTableRequest) returns (UpsertTableResponse);

  //-----External node management RPC interface
  rpc FetchDSSClusterConfig(google.protobuf.Empty)
      returns (FetchDSSClusterConfigResponse);
  rpc UpdateDSSClusterConfig(UpdateDSSClusterConfigRequest)
      returns (UpdateDSSClusterConfigResponse);
  // for migrate ds shard
  rpc ShardMigrate(ShardMigrateRequest) returns (ShardMigrateResponse);
  rpc ShardMigrateStatus(ShardMigrateStatusRequest)
      returns (ShardMigrateStatusResponse);
  rpc OpenDSShard(OpenDSShardRequest) returns (OpenDSShardResponse);
  rpc SwitchDSShardMode(SwitchDSShardModeRequest)
      returns (SwitchDSShardModeResponse);
  rpc UpdateDSShardConfig(UpdateDSShardConfigRequest)
      returns (UpdateDSShardConfigResponse);

  //-----New scan operations
  rpc ScanOpen(ScanRequest) returns (ScanResponse);
  rpc ScanNext(ScanRequest) returns (ScanResponse);
  rpc ScanClose(ScanRequest) returns (ScanResponse);

  //-----Create snapshot
  rpc CreateSnapshotForBackup(CreateSnapshotRequest)
      returns (CreateSnapshotResponse);

  //-----FaultInject for tests
  rpc FaultInjectForTest(FaultInjectRequest) returns (FaultInjectResponse);
};

enum DataStoreError {
  NO_ERROR = 0;
  DB_NOT_OPEN = 1;
  WRITE_TO_READ_ONLY_DB = 2;
  REQUESTED_NODE_NOT_OWNER = 3;
  WRITE_FAILED = 4;
  READ_FAILED = 5;
  TABLE_NOT_FOUND = 6;
  UPSERT_TABLE_FAILED = 7;
  UNKNOWN_ERROR = 8;
  BACKUP_FAILED = 9;
  STATUS_SWITCH_FAILED = 10;
}

enum CcTableType {
  Primary = 0;
  Secondary = 1;
  Catalog = 2;
  RangePartition = 3;
  UniqueSecondary = 4;
  RangeBucket = 5;
  ClusterConfig = 6;
}

enum KeyShardingErrorType {
  PrimaryNodeChanged = 0;
  NodeGroupChanged = 1;
}

message KeyShardingChanged {
  KeyShardingErrorType type = 1;
  DSSNodeBuf new_primary_node = 2;
  uint32 shard_id = 3;
  DSSClusterConfig new_cluster_config = 4;
}

message CommonResult {
  uint32 error_code = 1;
  string error_msg = 2;
  KeyShardingChanged new_key_sharding = 3;
}

message FetchRecordsRequest {
  message key {
    string kv_table_name_str = 1;
    CcTableType table_type = 2;
    string key_str = 3;
    uint32 key_shard_code = 4;
  }

  repeated key keys = 1;
  uint32 key_shard_code = 2;
}

message FetchRecordsResponse {
  message record {
    bytes payload = 1;
    uint64 version = 2;
    bool is_deleted = 3;
  }

  CommonResult result = 1;
  repeated record records = 2;
}

message BatchWriteRecordsRequest {
  message record {
    string key = 1;
    bytes value = 2;
    bool is_deleted = 3;
  }

  string kv_table_name = 1;
  uint32 key_shard_code = 2;
  repeated record records = 3;
}

message BatchWriteRecordsResponse {
  CommonResult result = 1;
}

message CkptEndRequest {
  string kv_table_name = 1;
  uint32 key_shard_code = 2;
}

message CkptEndResponse {
  CommonResult result = 1;
}

message FetchTableCatalogRequest {
  string table_name = 1;
  uint32 key_shard_code = 2;
}

message FetchTableCatalogResponse {
  CommonResult result = 1;
  bool found = 2;
  string schema_img = 3;
  uint64 version_ts = 4;
}

message ClusterNodeConfig {
  uint32 node_id = 1;
  string host_name = 2;
  int32 port = 3;
  bool is_candidate = 4;
}

message ClusterNodeGroupConfig {
  uint32 node_group_id = 1;
  repeated ClusterNodeConfig node_configs = 2;
}

message ReadClusterConfigRequest {
  uint32 key_shard_code = 1;
}

message ReadClusterConfigResponse {
  CommonResult result = 1;
  uint64 version = 2;
  bool uninitialized = 3;
  repeated ClusterNodeGroupConfig cluster_configs = 4;
}

message UpsertTableRequest {
  string table_name_str = 1;
  string old_schema_img = 2;
  string new_schema_img = 3;
  uint32 op_type = 4;
  uint64 commit_ts = 5;
  uint32 key_shard_id = 6;
}

enum UpsertTableOperationType {
  CreateTable = 0;
  DropTable = 1;
  TruncateTable = 2;
  AddIndex = 3;
  DropIndex = 4;
}

message UpsertTableResponse {
  CommonResult result = 1;
}

//-----External node management RPC interface

message DSSNodeBuf {
  string host_name = 1;
  int32 port = 2;
}

enum DSShardStatus {
  READ_ONLY = 0;
  READ_WRITE = 1;
  CLOSED = 2;
}

message DSShardBuf {
  uint32 shard_id = 1;
  uint64 shard_version = 2;
  repeated DSSNodeBuf member_nodes = 3;
}

message DSSClusterConfig {
  string sharding_algorithm = 1;
  repeated DSShardBuf shards = 2;
  uint64 topology_version = 3;
}

message FetchDSSClusterConfigResponse {
  DSSClusterConfig cluster_config = 1;
}

message UpdateDSSClusterConfigRequest {
  DSSClusterConfig cluster_config = 1;
}

message UpdateDSSClusterConfigResponse {
  uint32 error_code = 1;
  string error_msg = 2;
}

enum ShardMigrateError {
  IN_PROGRESS = 0;
  DUPLICATE_REQUEST = 1;
  REQUEST_NOT_OWNER = 2;
}

message ShardMigrateRequest {
  string event_id = 1;
  uint32 shard_id = 2;
  string to_node_host = 3;
  int32 to_node_port = 4;
}

message ShardMigrateResponse {
  string event_id = 1;
  ShardMigrateError error_code = 2;
}

message ShardMigrateStatusRequest {
  string event_id = 1;
}

message ShardMigrateStatusResponse {
  bool finished = 1;
  int32 status = 2;
}

message OpenDSShardRequest {
  uint32 shard_id = 1;
  DSShardStatus mode = 2;
  uint64 version = 3;
  repeated DSSNodeBuf members = 4;
}

message OpenDSShardResponse {
  uint32 error_code = 1;
  string error_msg = 2;
}

message SwitchDSShardModeRequest {
  uint32 shard_id = 1;
  DSShardStatus mode = 2;
  // for check
  uint64 version = 3;
}

message SwitchDSShardModeResponse {
  uint32 error_code = 1;
  string error_msg = 2;
}

message UpdateDSShardConfigRequest {
  uint32 shard_id = 1;
  uint64 version = 2;
  repeated DSSNodeBuf members = 3;
}

message UpdateDSShardConfigResponse {
  uint32 error_code = 1;
  string error_msg = 2;
}

//-----New scan operations

// Search condition for pushdown filtering
message SearchCondition {
  string field_name = 1;
  string op = 2;     // Operation type (e.g., EQ, LT, GT, etc.)
  string value = 3;  // Serialized value to compare against
}

// Request message for scan operations
message ScanRequest {
  uint32 req_shard_id = 1;
  string session_id = 2;
  string kv_table_name_str = 3;
  // it is true if start key is neg infinity
  bool start_key_is_neg_inf = 4;
  // it is true if start key is pos infinity
  bool start_key_is_pos_inf = 5;
  bytes start_key = 6;
  bool inclusive = 7;     // Whether to include start_key in results
  bool scan_forward = 8;  // Scan direction
  uint32 batch_size = 9;  // Number of items to return in scan
  repeated SearchCondition search_conditions = 10;
}

// Response message for scan operations
message ScanResponse {
  message Item {
    bytes key = 1;
    bytes value = 2;
    // decode will happen on client side for now,
    // although it can be more efficent
    // if we can do filtering and ttl checking on server side
    // uint64 version = 3;
    // bool deleted = 4;
  }

  CommonResult result = 1;
  repeated Item items = 2;
  string session_id = 3;
}

message CreateSnapshotRequest {
  uint32 req_shard_id = 1;
  string backup_name = 2;
}

message CreateSnapshotResponse {
  repeated string snapshot_files = 1;
  CommonResult result = 2;
}

//-----FaultInject for test
message FaultInjectRequest {
  string fault_name = 1;
  string fault_paras =
      2;  // eg. "action=NOTIFY_CHECKPOINTER", "remove", "action=panic".
}
message FaultInjectResponse {
  bool finished = 1;
}
